File: merge_sort.cpp
#include <iostream>

using namespace std;

int* mergeSort(int[], int);
void printArray(int[], int);

int main() {
    int N;
    
    cout << "Merge Sort" << endl;
    cout << "Ingresa el número de elementos de la lista: ";
    cin >> N;
    
    int nums[N];
    int* numsOrd;
    
    for(int i = 0; i < N; i++) {
        cin >> nums[i]; 
    }
    
    cout << "Arreglo ordenado" << endl;
    numsOrd = mergeSort(nums, N);
    printArray(numsOrd, N);
    
    delete[] numsOrd;
    return 0;
}

int* mergeSort(int arr[], int size){
    if(size == 1){
        return arr;
    }
    
    int mid = size / 2;
    int izquierdo[mid];
    int derecho[size - mid];
    int j = 0;
    
    for(int i = 0; i < size; i++) {
        if(i < mid){
            izquierdo[i] = arr[i];
        }else {
            derecho[j] = arr[i];
            j++;
        }
    }
    
    int* izquierdo_ordenado = mergeSort(izquierdo, mid);
    int* derecho_ordenado = mergeSort(derecho, size - mid);
    
    int i = 0, k = 0;
    j = 0;
    
    int* ordenado = new int[size];
    
    while(i < mid && j < (size - mid)){
        if(izquierdo_ordenado[i] < derecho_ordenado[j]){
            ordenado[k] = izquierdo_ordenado[i];
            i++;
        }else {
            ordenado[k] = derecho_ordenado[j];
            j++;
        }
        k++;
    }
    
    while(i < mid){
        ordenado[k] = izquierdo_ordenado[i];
        i++;
        k++;
    }
    
    while(j < (size - mid)){
        ordenado[k] = derecho_ordenado[j];
        j++;
        k++;
    }
    
    return ordenado;
}

void printArray(int arr[], int size){
    for(int i = 0; i < size; i++) {
        cout << arr[i] << endl;
    }
}

File: merge_sort.py
def merge_sort(lista):
    if(len(lista) == 1 ):
        return lista
    
    mid = len(lista) // 2

    izquierdo = lista[:mid]
    derecho = lista[mid:]

    izquierdo_ordenado = merge_sort(izquierdo)
    derecho_ordenado = merge_sort(derecho)

    merged = []

    i, j = 0, 0

    while(i < mid and j < (len(lista) - mid)):
        if(izquierdo_ordenado[i] < derecho_ordenado[j]):
            merged.append(izquierdo_ordenado[i]) 
            i += 1
        else:
            merged.append(derecho_ordenado[j])
            j += 1

    merged += izquierdo_ordenado[i:]
    merged += derecho_ordenado[j:]

    return merged


print("Merge sort\n")
lista = [int(x) for x in input("Ingresa una lista: ").split(" ")]
print("\nLista original: ", lista)
print("Lista ordenada: ", merge_sort(lista))